File: pas/tio.pas
Unit TIO;

{$mode objfpc} // Modern Object Pascal

Interface

Uses
{$IFDEF LINUX}
	BaseUnix, TermIO,
{$ENDIF}
{$IFDEF WINDOWS}
	Windows,
{$ENDIF}
	SysUtils;

Type
	PTerminalState = ^TTerminalState;
	TTerminalState = Record
		{$IFDEF WIN32}
		Input: THandle;
		Output: THandle;
		OriginalInputMode: DWORD;
		OriginalOutputMode: DWORD;
		{$ELSE}
		OriginalTermios: termios;
		{$ENDIF}
		IsRaw: Boolean;
	End;

Procedure TerminalStateInit(State: PTerminalState);
Procedure InitTerminal(State: PTerminalState);
Procedure RestoreTerminal(State: PTerminalState);
Procedure GotoXY(X, Y: Integer);
Procedure CursorOn;
Procedure CursorOff;
Procedure ClearScreen;
Function TerminalReadChar(State: PTerminalState): Integer;
Function GetChar(State: PTerminalState): Integer;

Implementation

// -------------------------
// Terminal State Management
// -------------------------

Procedure TerminalStateInit(State: PTerminalState);
Begin
	{$IFDEF WIN32}
	State^.Input := INVALID_HANDLE_VALUE;
	State^.OriginalInputMode := 0;
	State^.Output := INVALID_HANDLE_VALUE;
	State^.OriginalOutputMode := 0;
	{$ENDIF}
	State^.IsRaw := False;
End;

Procedure InitTerminal(State: PTerminalState);
Var
{$IFDEF WIN32}
	Mode: DWORD;
{$ELSE}
	Raw: termios;
{$ENDIF}
Begin
	If (State = nil) Or State^.IsRaw Then
		Exit;

	{$IFDEF WIN32}
	State^.Input := GetStdHandle(STD_INPUT_HANDLE);
	If State^.Input = INVALID_HANDLE_VALUE Then
	Begin
		WriteLn(StdErr, 'Error: GetStdHandle(STD_INPUT_HANDLE) failed');
		Exit;
	End;

	State^.Output := GetStdHandle(STD_OUTPUT_HANDLE);
	If State^.Output = INVALID_HANDLE_VALUE Then
	Begin
		WriteLn(StdErr, 'Error: GetStdHandle(STD_OUTPUT_HANDLE) failed');
		Exit;
	End;

	If Not GetConsoleMode(State^.Input, @State^.OriginalInputMode) Then
	Begin
		WriteLn(StdErr, 'Error: GetConsoleMode failed for stdin');
		Exit;
	End;

	Mode := State^.OriginalInputMode;
	Mode := Mode Or ENABLE_VIRTUAL_TERMINAL_INPUT;
	Mode := Mode And Not ENABLE_PROCESSED_INPUT;
	Mode := Mode And Not ENABLE_LINE_INPUT;
	Mode := Mode And Not ENABLE_ECHO_INPUT;

	If Not SetConsoleMode(State^.Input, Mode) Then
		Exit;

	If Not GetConsoleMode(State^.Output, @State^.OriginalOutputMode) Then
	Begin
		WriteLn(StdErr, 'Error: GetConsoleMode failed for stdout');
		Exit;
	End;

	Mode := State^.OriginalOutputMode;
	Mode := Mode Or ENABLE_VIRTUAL_TERMINAL_PROCESSING;
	If Not SetConsoleMode(State^.Output, Mode) Then
		Exit;
	{$ELSE}
	If TCGetAttr(0, State^.OriginalTermios) < 0 Then
	Begin
		WriteLn(StdErr, 'Error: TCGetAttr failed');
		Exit;
	End;
	Raw := State^.OriginalTermios;
	Raw.c_lflag := Raw.c_lflag And Not (ICANON Or ECHO Or ISIG);
	Raw.c_iflag := Raw.c_iflag And Not (IXON Or ICRNL);
	Raw.c_cc[VMIN] := 1;
	Raw.c_cc[VTIME] := 0;

	If TCSetAttr(0, TCSANOW, raw) < 0 Then
	Begin
		WriteLn(StdErr, 'Error: TCSetAttr failed');
		Exit;
	End;
	{$ENDIF}

	State^.IsRaw := True;
End;

Procedure RestoreTerminal(State: PTerminalState);
Begin
	If (State = nil) Or Not State^.IsRaw Then
		Exit;

	{$IFDEF WIN32}
	If State^.Input <> INVALID_HANDLE_VALUE Then
		SetConsoleMode(State^.Input, State^.OriginalInputMode);
	If State^.Output <> INVALID_HANDLE_VALUE Then
		SetConsoleMode(State^.Output, State^.OriginalOutputMode);
	{$ELSE}
	TCSetAttr(0, TCSANOW, State^.OriginalTermios);
	{$ENDIF}

	State^.IsRaw := False;
End;

// -------------------------
// Terminal Control Functions
// -------------------------

Procedure GotoXY(X, Y: Integer);
Begin
	Write(Format(#27'[%d;%dH', [Y, X]));
	Flush(Output);
End;

Procedure CursorOn;
Begin
	Write(#27'[?25h');
	Flush(Output);
End;

Procedure CursorOff;
Begin
	Write(#27'[?25l');
	Flush(Output);
End;

Procedure ClearScreen;
Begin
	Write(#27'[2J'#27'[H');
	Flush(Output);
End;

// -------------------------
// Input Handling
// -------------------------

Function TerminalReadChar(State: PTerminalState): Integer;
Var
		Ch: Char;
	{$IFDEF WIN32}
		InputRecord: TInputRecord;
		EventsRead: DWORD;
		CharsRead: DWORD;
	{$ELSE}
		ReadFds: TFDSet;
		TimeOut: TTimeVal;
	{$ENDIF}
Begin
	Result := -1;
	If (State = nil) Or Not State^.IsRaw Then
		Exit;

	{$IFDEF WIN32}
	While True Do
	Begin
		If Not PeekConsoleInput(State^.Input, InputRecord, 1, EventsRead) Then
			Exit;

		If EventsRead = 0 Then
			Exit;

		If (InputRecord.EventType = KEY_EVENT) And 
			(InputRecord.Event.KeyEvent.bKeyDown) Then
			Break;

		ReadConsoleInput(State^.Input, InputRecord, 1, EventsRead);
	End;

	If Not ReadConsoleA(State^.Input, @Ch, 1, CharsRead, nil) Then
		Exit;
	Result := Ord(Ch);
	{$ELSE}
	TimeOut.tv_sec := 0;
	TimeOut.tv_usec := 0;

	fpFD_ZERO(ReadFds);
	fpFD_SET(StdInputHandle, ReadFds);
	
	If (fpSelect(StdInputHandle + 1, @ReadFds, nil, nil, @TimeOut) > 0) And 
		(fpFD_ISSET(StdInputHandle, ReadFds) > 0) Then
	Begin
		If fpRead(StdInputHandle, @Ch, 1) <> 1 Then
			Exit;
	End
	Else
		Exit;
	{$ENDIF}
End;

Function GetChar(State: PTerminalState): Integer;
type
	TParserState = (
		STATE_NORMAL,
		STATE_ESC,
		STATE_CSI,
		STATE_OSC,
		STATE_SS3,
		STATE_IGNORE
	);
Var
	ParserState: TParserState;
	OscEsc: Boolean;
	Ch: Integer;
Begin
	ParserState := STATE_NORMAL;
	OscEsc := False;

	while True Do
	Begin
		Ch := TerminalReadChar(State);
		If Ch = -1 Then
		Begin
			Result := -1;
			Exit;
		End;

		Case ParserState Of
			STATE_NORMAL:
				If Ch = 27 Then
					ParserState := STATE_ESC
				Else
					Break;

			STATE_ESC:
				If Ch = Ord('[') Then
					ParserState := STATE_CSI
				Else If Ch = Ord(']') Then
				Begin
					ParserState := STATE_OSC;
					OscEsc := False;
				end
				Else If Ch = Ord('O') Then
					ParserState := STATE_SS3
				Else
					Break;

			STATE_CSI:
				If (Ch >= $40) And (Ch <= $7E) Then
					ParserState := STATE_NORMAL;

			STATE_OSC:
				Begin
					If OscEsc Then
					Begin
						If Ch = Ord('\') Then
							ParserState := STATE_NORMAL;
						OscEsc := False;
					end
					Else If Ch = 27 Then
						OscEsc := True
					Else If Ch = Ord(#7) Then // BEL
						ParserState := STATE_NORMAL;
				End;

			STATE_SS3:
				ParserState := STATE_NORMAL;

			STATE_IGNORE:
				If (Ch >= 32) And (Ch <= 126) Then
					ParserState := STATE_NORMAL;
		End;
	End;

	Result := Ch;
End;

End.

File: src/tio.c
#include "tio.h"

#ifndef _WIN32
	#include <unistd.h> // STDIN_FILENO
	#include <sys/select.h> // select
	#include <sys/time.h> // timeval
#endif
#include <stdio.h>  //printf
#include <stdlib.h> // malloc, free

struct terminal_state {
#ifdef _WIN32
	HANDLE input;
	HANDLE output;
	DWORD original_input_mode;
	DWORD original_output_mode;
#else
	struct termios original_termios;
#endif
};

static TERM* terminal_state_init(void) {
	TERM* term = malloc(sizeof(TERM));
	if (term == NULL)
		return NULL;
	
#ifdef _WIN32
	term->input = INVALID_HANDLE_VALUE;
	term->output = INVALID_HANDLE_VALUE;
	term->original_input_mode = 0;
	term->original_output_mode = 0;
#else
	tcgetattr(STDIN_FILENO, &term->original_termios);
#endif
	return term;
}

TERM* init_terminal(void) {
	TERM* term = terminal_state_init();
	if (term == NULL)
		return NULL;

#ifdef _WIN32
	term->input = GetStdHandle(STD_INPUT_HANDLE);
	if (term->input == INVALID_HANDLE_VALUE) {
		fprintf(stderr, "Error: GetStdHandle(STD_INPUT_HANDLE) failed\n");
		return NULL;
	}

	term->output = GetStdHandle(STD_OUTPUT_HANDLE);
	if (term->output == INVALID_HANDLE_VALUE) {
		fprintf(stderr, "Error: GetStdHandle(STD_OUTPUT_HANDLE) failed\n");
		return NULL;
	}

	// Save the original mode
	if (!GetConsoleMode(term->input, &term->original_input_mode)) {
		fprintf(stderr, "Error: GetConsoleMode failed for stdin\n");
		return NULL;
	}

	DWORD mode = term->original_input_mode;

	// Key settings for VT input:
	mode |= ENABLE_VIRTUAL_TERMINAL_INPUT;  // Enable VT sequences
	mode &= ~ENABLE_PROCESSED_INPUT;        // Get "raw" escape codes
	mode &= ~ENABLE_LINE_INPUT;             // Non-canonical mode
	mode &= ~ENABLE_ECHO_INPUT;             // No echo

	// Optional for mouse
	// mode |= ENABLE_MOUSE_INPUT;

	if (!SetConsoleMode(term->input, mode)) {
		// Try without ENABLE_VIRTUAL_TERMINAL_INPUT (for older Windows)
		//mode &= ~ENABLE_VIRTUAL_TERMINAL_INPUT;
		//if (!SetConsoleMode(term->input, mode)) {
		//	fprintf(stderr, "Error: SetConsoleMode failed\n");
			return NULL;
		//}
	}

	if (!GetConsoleMode(term->output, &term->original_output_mode)) {
		fprintf(stderr, "Error: GetConsoleMode failed for stdout\n");
		return NULL;
	}

	mode = term->original_output_mode;
	mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
	if (!SetConsoleMode(term->output, mode))
		return NULL;
#else
	// Save current settings
	if (tcgetattr(STDIN_FILENO, &term->original_termios) == -1) {
		fprintf(stderr, "tcgetattr failed\n");
		return NULL;
	}
	
	// Set non-canonical mode
	struct termios raw = term->original_termios;
	raw.c_lflag &= ~(ICANON | ECHO);
	// Additional settings for more "raw" mode:
	raw.c_lflag &= ~ISIG;  // Disable signal handling (Ctrl+C, Ctrl+Z)
	raw.c_iflag &= ~(IXON | ICRNL); // Disable flow control and CR->NL conversion
	
	// Configure minimum number of characters and timeout
	raw.c_cc[VMIN] = 1;   // Read at least 1 character
	raw.c_cc[VTIME] = 0;  // No timeout
	
	if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw) == -1) {
		fprintf(stderr, "tcsetattr failed\n");
		return NULL;
	}
#endif
	return term;
}

void restore_terminal(TERM* term) {
	if (term == NULL)
		return;

#ifdef _WIN32
	if (term->input == INVALID_HANDLE_VALUE)
		return;
	
	SetConsoleMode(term->input, term->original_input_mode);

	if (term->output == INVALID_HANDLE_VALUE)
		return;

	SetConsoleMode(term->output, term->original_output_mode);
#else
   
	tcsetattr(STDIN_FILENO, TCSAFLUSH, &term->original_termios);
#endif
	free(term);
}

void goto_xy(int x, int y) {
    printf("\033[%d;%dH", y, x);
    fflush(stdout);
}

void cursor_on(void) {
	const char esc_seq[] = "\033[?25h";
	fwrite(esc_seq, 1, sizeof(esc_seq) - 1, stdout);
	fflush(stdout);
}

void cursor_off(void) {
	const char esc_seq[] = "\033[?25l";
	fwrite(esc_seq, 1, sizeof(esc_seq) - 1, stdout);
	fflush(stdout);
}

void clear_screen(void) {
	const char esc_seq[] = "\033[2J\033[H";
	fwrite(esc_seq, 1, sizeof(esc_seq) - 1, stdout);
	fflush(stdout);
}

int terminal_read_char(TERM* term) {
	if (term == NULL)
		return EOF;

	char ch;
#ifdef _WIN32
	while (1) {
		INPUT_RECORD input_record;
		DWORD events_read = 0;

		if (!PeekConsoleInput(term->input, &input_record, 1, &events_read))
			return EOF;

		if (events_read == 0)
			return EOF;

		if (input_record.EventType == KEY_EVENT &&
			input_record.Event.KeyEvent.bKeyDown)
			break;

		ReadConsoleInput(term->input, &input_record, 1, &events_read);
	}

	DWORD charsRead = 0;
	if (!ReadConsoleA(term->input, &ch, 1, &charsRead, NULL))
		return EOF;
#else
	fd_set readfds;
    struct timeval timeout = {0, 0};
    
    FD_ZERO(&readfds);
    FD_SET(STDIN_FILENO, &readfds);
    
    int result = select(STDIN_FILENO + 1, &readfds, NULL, NULL, &timeout);
    
    if (result > 0 && FD_ISSET(STDIN_FILENO, &readfds) && read(STDIN_FILENO, &ch, 1) != 1)
		return EOF;
#endif
	return (unsigned char)ch;
}
